src/protocol/malicious.rs

```diff
@@ -0,0 +1,350 @@
+    }
+
+    #[allow(dead_code)]
+    pub async fn validate(&self) -> Result<(), BoxError> {
+        let record_3 = RecordId::from(3);
+        let record_4 = RecordId::from(4);
+        let record_5 = RecordId::from(5);
+        let record_6 = RecordId::from(6);
+
+        // send our `u_i+1` value to the helper on the right
+        let channel = self.ctx.mesh();
+        try_join(
+            channel.send(
+                self.ctx.role().peer(Direction::Right),
+                record_3,
+                UValue { payload: self.u },
+            ),
+            channel.send(
+                self.ctx.role().peer(Direction::Right),
+                record_4,
+                UValue { payload: self.w },
+            ),
+        )
+        .await?;
+
+        // receive `u_i` value from helper to the left
+        let (u_left_struct, w_left_struct): (UValue<F>, UValue<F>) = try_join(
+            channel.receive(self.ctx.role().peer(Direction::Left), record_3),
+            channel.receive(self.ctx.role().peer(Direction::Left), record_4),
+        )
+        .await?;
+
+        let u_left = u_left_struct.payload;
+        let w_left = w_left_struct.payload;
+
+        let u_share = Replicated::new(u_left, self.u);
+        let w_share = Replicated::new(w_left, self.w);
+
+        // This should probably be done in parallel with the futures above
+        let r = reveal(self.ctx.narrow(&Step::RevealR), record_5, self.r_share).await?;
+        let t = u_share - (w_share * r);
+
+        let is_valid = check_zero(self.ctx.narrow(&Step::CheckZero), record_6, t).await?;
+
+        if is_valid {
+            Ok(())
+        } else {
+            Err(Box::new(Error::MaliciousSecurityCheckFailed))
+        }
+    }
+}
+
+#[cfg(test)]
+pub mod tests {
+    use crate::error::BoxError;
+    use crate::field::Fp31;
+    use crate::protocol::{
+        malicious::{SecurityValidator, Step},
+        QueryId, RecordId,
+    };
+    use crate::test_fixture::{logging, make_contexts, make_world, share, TestWorld};
+    use futures::future::{try_join, try_join_all};
+    use proptest::prelude::Rng;
+
+    #[tokio::test]
+    async fn test_simplest_circuit() -> Result<(), BoxError> {
+        logging::setup();
+
+        let world: TestWorld = make_world(QueryId);
+        let context = make_contexts(&world);
+        let mut rng = rand::thread_rng();
+
+        let a = Fp31::from(rng.gen::<u128>());
+        let b = Fp31::from(rng.gen::<u128>());
+
+        let a_shares = share(a, &mut rng);
+        let b_shares = share(b, &mut rng);
+
+        let futures = (0..3).into_iter().zip(context).map(|(i, ctx)| async move {
+            let mut v =
+                SecurityValidator::new(ctx.narrow(&"SecurityValidatorInternal".to_string()));
+            let r_share = v.r_share();
+
+            let a_record_narrowed_context = ctx.narrow(&"record_1_hack".to_string());
+            let b_record_narrowed_context = ctx.narrow(&"record_2_hack".to_string());
```

Comment:
I think that you just want:

```rust
a_ctx = ctx.narrow("multiply");
b_ctx = ctx.clone();
```

Rather than having each narrow separately.

Then you use `narrow()` each of those for each of:

* a*r or b*r
* accumulate_macs
* the multiply of a*b
* the multiply of a*(b*r) or b*(a*r)
* the multiply of r*(a*b)